src/main/java/net/helix/pendulum/controllers/RoundViewModel.java

```diff
@@ -474,10 +528,45 @@ public Integer index() {
         return round.index.getValue();
     }
 
+    /**
+     * Return merkle root for accepted milestone from this round.
+     *
+     * @return hash of the merkle root.
+     */
     public Hash getMerkleRoot() {
-        List<List<Hash>> merkleTree = Merkle.buildMerkleTree(new LinkedList<>(getHashes()));
-        Hash root = merkleTree.get(merkleTree.size()-1).get(0);
-        return root;
+        MerkleTree merkle = new MerkleTreeImpl();
+        return HashFactory.TRANSACTION.create(merkle.getMerkleRoot(new LinkedList<>(getHashes()), MerkleOptions.getDefault()));
+    }
+
+    public Hash getMerkleRoot(Tangle tangle, SnapshotProvider snapshotProvider, LedgerService ledgerService) {
+        List<Hash> nonConflictualMilestones =  getMilestonesToApprove(tangle, snapshotProvider, ledgerService);
+        return HashFactory.TRANSACTION.create( new MerkleTreeImpl().getMerkleRoot(new LinkedList<>(nonConflictualMilestones), MerkleOptions.getDefault()));
+    }
+
+    public List<Hash> getMilestonesToApprove(Tangle tangle, SnapshotProvider snapshotProvider, LedgerService ledgerService) {
+
+        List<TransactionViewModel> targetRoundMilestones = TransactionViewModel.fromHashes(getHashes(), tangle);
+        Snapshot snapshot = snapshotProvider.getInitialSnapshot();
+
+        if (snapshot.isConsistent()) {
+            try {
+                Set<Hash> approvedMilestones = new HashSet<Hash>();
+                ledgerService.getSnapshotService().replayMilestones(snapshot, index() - 1);
+                ledgerService.applyRoundToLedger(this, snapshot, approvedMilestones, true);
+                targetRoundMilestones.stream().filter(m -> !approvedMilestones.contains(m)).forEach(m -> {
+                    removeMilestone(m.getHash());
+                });
+                update(tangle);
+            } catch (LedgerException e) {
```

Comment:
- one catch is enough here
- grammar: simultaneously

Reply:
done